server/src/main/java/org/elasticsearch/search/retriever/rankdoc/RankDocsQuery.java

```diff
@@ -33,142 +40,307 @@
+                                    return slowAdvance(target);
+                                }
+
+                                @Override
+                                public long cost() {
+                                    return upper - lower;
+                                }
+                            };
+                        }
+
+                        @Override
+                        public float getMaxScore(int docId) {
+                            if (docId != NO_MORE_DOCS) {
+                                docId += context.docBase;
+                            }
+                            float maxScore = 0;
+                            for (int idx = Math.max(lower, upTo); idx < upper && docs[idx].doc <= docId; idx++) {
+                                maxScore = Math.max(maxScore, docs[idx].score);
+                            }
+                            return maxScore;
+                        }
+
+                        @Override
+                        public float score() {
+                            return docs[upTo].score;
+                        }
+
+                        @Override
+                        public int docID() {
+                            return currentDocId();
+                        }
+
+                        private int currentDocId() {
+                            if (upTo == -1) {
+                                return -1;
+                            }
+                            if (upTo >= upper) {
+                                return NO_MORE_DOCS;
+                            }
+                            return docs[upTo].doc - context.docBase;
+                        }
+
+                    };
+                }
+
+                @Override
+                public boolean isCacheable(LeafReaderContext ctx) {
+                    return true;
+                }
+            };
+        }
+
+        @Override
+        public String toString(String field) {
+            return this.getClass().getSimpleName() + "{rank_docs:" + Arrays.toString(docs) + "}";
+        }
+
+        @Override
+        public void visit(QueryVisitor visitor) {
+            visitor.visitLeaf(this);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (sameClassAs(obj) == false) {
+                return false;
+            }
+            TopQuery other = (TopQuery) obj;
+            return Arrays.equals(docs, other.docs)
+                && Arrays.equals(segmentStarts, other.segmentStarts)
+                && contextIdentity == other.contextIdentity;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(classHash(), Arrays.hashCode(docs), Arrays.hashCode(segmentStarts), contextIdentity);
+        }
+    }
 
     private final RankDoc[] docs;
-    private final int[] segmentStarts;
-    private final Object contextIdentity;
+    private final Query topQuery;
+    private final Query tailQuery;
```

Comment:
Could you add some java comments to the internal parameter. I honestly don't know the purpose of tailQuery. One would assume that at this point rankDocs is the results of previously running `sources` queries.

Reply:
Added some initial documentation in b623bee80793b34086ad221acaea4c881511c69f